Update distributing components section with docs on wkg (#220)
bytecodealliance/component-docs: component-model/src/creating-and-consuming/distributing.md

You can get involved with improving the packaging and hosting of Wasm components by joining the [Bytecode Alliance Packaging Special Interest Group (SIG)](https://github.com/bytecodealliance/governance/blob/main/SIGs/sig-packaging/proposal.md).

## Registry Tooling

The [`wasm-pkg-tools` project](https://github.com/bytecodealliance/wasm-pkg-tools) enables fetching and publishing Wasm Components to OCI or Warg registries. It contains a `wkg` CLI tool that eases distributing and fetching components and WIT packages. The following examples walk through using `wkg`.

## Distributing Components Using `wkg`

A component is pushed to an OCI registry using `wkg oci push`. The example below pushes to a GHCR:

```sh
wkg oci push ghcr.io/user/hello:0.1.0 hello.wasm
```

## Fetching Components Using `wkg`

A component is pulled from a OCI registry using `wkg oci pull`. The example below pulls a component from GHCR:

```sh
wkg oci pull ghcr.io/user/hello:0.1.0 -o hello.wasm
```

## Configuring `wkg` to Fetch WASI Packages

> The `wkg` tool uses a configuration file to store settings with a default location of `$XDG_CONFIG_HOME/wasm-pkg/config.toml`. It must be configured to know which registry to use for which package namespaces. The following is a convenient configuration to ensure you can fetch WASI packages from the [WebAssembly OCI registry](https://github.com/WebAssembly/WASI/pkgs/container/wasi), where the latest interfaces are published upon WASI releases.

Reviewer: I see there is a `wkg config --edit` which might be more convenient than finding the file by hand.

```toml
# $XDG_CONFIG_HOME/wasm-pkg/config.toml
default_registry = "ghcr.io"

[namespace_registries]
wasi = { registry = "wasi",  metadata = { preferredProtocol = "oci", "oci" = {registry = "ghcr.io", namespacePrefix = "webassembly/" } } }

[package_registry_overrides]

[registry]
```

## Distributing WIT Packages using  `wkg`

The `wkg` tool uses a [configuration file](https://github.com/bytecodealliance/wasm-pkg-tools?tab=readme-ov-file#configuration) to store settings with a default location of `$XDG_CONFIG_HOME/wasm-pkg/config.toml`. It must be configured to know which registry to use for which package namespaces. The following configuration, instructs `wkg` to use [ttl.sh](https://ttl.sh/) OCI registry for all packages with the `foo` namespace.

```toml
# $XDG_CONFIG_HOME/wasm-pkg/config.toml
default_registry = "ghcr.io"

[namespace_registries]

> foo = { registry = "foo-registry",  metadata = { preferredProtocol = "oci", "oci" = {registry = "ttl.sh", namespacePrefix = "wasm-components/" } } }

Author: it can be referenced to do package overrides in another section of the config. I didn't think it was necessary to go to that level of detail, though

[package_registry_overrides]

[registry]
```

Now, `foo` packages can be built and published using `wkg`.

```sh
mkdir wit
cat > wit/foo.wit << EOL

> package foo:bar@0.1.0;

Author: switched to using the adder interface from the tutorial

interface do-something {
    reduce: func(a: u32, b: u32) -> u32;
}

world example {
    export do-something;
}
EOL

wkg wit build

> wkg publish foo:bar@0.1.0.wasm

Reviewer: When should I use `publish` vs `oci push`?

> wkg publish foo:bar@0.1.0.wasm

Reviewer: I really like how you present the "surprise .wasm file" now - thanks!

Publishing and distribution are not defined by the core component model, but form important part of the component ecosystem. For example, if you're writing JavaScript, and want to pull in a highly optimised machine learning algorithm written in C and compiled to Wasm, you can pull it from a registry, ideally just as easily as you would add a NPM package from the NPM registry.

You can get involved with improving the packaging and hosting of Wasm components by joining the [Bytecode Alliance Packaging Special Interest Group (SIG)](https://github.com/bytecodealliance/governance/blob/main/SIGs/sig-packaging/proposal.md).

## The `wkg` Registry Tool

The [`wasm-pkg-tools` project](https://github.com/bytecodealliance/wasm-pkg-tools) enables fetching and publishing Wasm components to OCI registries. It contains a `wkg` CLI tool that eases distributing and fetching components and WIT packages. `wkg` contains several subcommand:

- `wkg oci` - is a CLI wrapper around the [oci-wasm](https://github.com/bytecodealliance/rust-oci-wasm) crate which enables pushing/pulling Wasm artifacts to/from any OCI registry

> - `wkg publish` - pushes *library* components or WIT packages

Author: I wanted to emphasize that wkg publish is used for publishing components that are meant to be consumed rather than executed, but i guess you could publish a component that implements `wasi:cli/run`.

> - `wkg publish` - pushes *library* components or WIT packages

Author: This is a good distinction of "by package name". I incorporated your suggestions

- `wkg get` - pulls *library* components or WIT packages
- `wkg wit` - commands for interacting with WIT files and dependencies
- `wkg config` - interact with the `wkg` configuration

The following sections detail a subset of actions that can be performed with `wkg`.

## `wkg` Configuration Files

> The `wkg` tool uses a configuration file to understand where to publish and fetch specific packages to and from. It provides the ability to configure:

Author: Added this line about mapping package name to location. I like it!

> - a default registry for all packages at the top level of the file

Author: I like these! Updated.

Imagine you have defined the following `adder` world in WIT:

```wit
package docs:adder@0.1.0;

interface add {
    add: func(a: u32, b: u32) -> u32;
}

world adder {
    export add;
}
```

You can publish this *WIT* using `wkg` by wrapping it up as a Wasm component. Yes, you heard that right! We are packaging WIT as Wasm.

```sh
# Package the contents of add WIT directory as Wasm
wkg wit build --wit-dir tutorial/wit/adder
# Publish the produced component
wkg publish docs:adder@0.1.0.wasm
```

If you had configured `wkg` as described in the [`wkg` configuration section](#wkg-configuration-files), this would publish the component to `ttl.sh/wasm-components/docs/adder:0.1.0`. This WIT can then be fetched using `wkg get`, specifying the format `wit`:

```sh
wkg get --format wit docs:adder@0.1.0
```

Instead of publishing the WIT interface, you could publish the built component by running:

```sh
wkg publish adder.wasm --package docs:adder@0.1.0
```

This component can then be fetched by running:

```sh
wkg get docs:adder@0.1.0
```

## More Generic Operations with `wkg oci`

> The `wkg oci` subcommand is a CLI wrapper around the [oci-wasm](https://github.com/bytecodealliance/rust-oci-wasm) crate which enables pushing/pulling Wasm artifacts to/from any OCI registry. Unlike `wkg publish` and `wkg get` it is not limited to library components, as providing the WIT package is not required.

Author: I removed all reference to library components and instead used the distinction of whether or not the artifact is being referenced by package name

> A component is pushed to an OCI registry using `wkg oci pull`. The example below pulls a component from a GitHub Container Registry.

Reviewer: ```suggestion
To push a component to an OCI registry, use `wkg oci pull`. The example below pushes a component to a GitHub Container Registry.
```

Publishing and distribution are not defined by the core component model, but form important part of the component ecosystem. For example, if you're writing JavaScript, and want to pull in a highly optimised machine learning algorithm written in C and compiled to Wasm, you can pull it from a registry, ideally just as easily as you would add a NPM package from the NPM registry.

You can get involved with improving the packaging and hosting of Wasm components by joining the [Bytecode Alliance Packaging Special Interest Group (SIG)](https://github.com/bytecodealliance/governance/blob/main/SIGs/sig-packaging/proposal.md).

## The `wkg` Registry Tool

> The [`wasm-pkg-tools` project](https://github.com/bytecodealliance/wasm-pkg-tools) enables fetching and publishing Wasm components to OCI registries. It contains a `wkg` CLI tool that eases distributing and fetching components and WIT packages. The usual way of using `wkg` is to address packages by their package name, i.e. `example:adder@1.0.0`. When using `wkg` this way, you don't need to know about the physical location of the package , as the `wkg` configuration handles that for you. If you need to, though, you can also use `wkg` to work with OCI artifacts directly, addressing them by OCI references when using the `wkg oci` subcommand.

Reviewer: ```suggestion
The [`wasm-pkg-tools` project](https://github.com/bytecodealliance/wasm-pkg-tools) enables fetching and publishing Wasm components to OCI registries. It contains a `wkg` CLI tool that eases distributing and fetching components and WIT packages. The usual way of using `wkg` is to address packages by their package name, i.e. `example:adder@1.0.0`. When using `wkg` this way, you don't need to know about the physical location of the package, as the `wkg` configuration handles that for you. If you need to, though, you can also use `wkg` to work with OCI artifacts directly, addressing them by OCI references when using the `wkg oci` subcommand.
```

- Registry overrides for specific packages, or packages not stored in the same place as the rest of their namespace. For example, if `wasi:key-value` were stored in a different registry from other `wasi` packages.  
- The default registry for all packages not listed in one of the previous sections  

The configuration file also includes credentials for private registries, or for pushing to registries where you have permission, and other configuration options.  See the [`wkg` docs for more configuration options](https://github.com/bytecodealliance/wasm-pkg-tools?tab=readme-ov-file#configuration).

For example, to fetch WASI packages, such as `wasi:clocks` and `wasi:http`, you can add a line under the `namespace_registries` section for the `wasi` namespace. Specifically, the example below configures `wkg` to fetch WASI packages from the [WebAssembly OCI GitHub Container Registry](https://github.com/orgs/WebAssembly/packages), where the latest interfaces are published upon WASI releases. To edit your `wkg` config file, run `wkg config --edit`. 

> > Remember, all package names consist of the a namespace field followed by the package field. The package name `wasi:clocks` has a namespace of `wasi` and package field of `clocks`. In this way, the following configuration ensures that `wkg` will know to route fetches and publishes of any `wasi:x` to the configured location.

Reviewer 2: nit: Maybe put a variable instead of just `x` in `wasi:x` for clarity: e.g. `wasi:<package_name>` or something like that

```toml
# $XDG_CONFIG_HOME/wasm-pkg/config.toml
default_registry = "ghcr.io"

[namespace_registries]
# Tell wkg that packages with the `wasi` namespace are in an OCI registry under ghcr.io/webassembly 
wasi = { registry = "wasi",  metadata = { preferredProtocol = "oci", "oci" = {registry = "ghcr.io", namespacePrefix = "webassembly/" } } }
```

As a more generic example, The following configuration, instructs `wkg` to use [ttl.sh](https://ttl.sh/) OCI registry for all packages with the `docs` namespace.

```toml
# $XDG_CONFIG_HOME/wasm-pkg/config.toml
default_registry = "ghcr.io"

[namespace_registries]
# Instruct wkg to use the OCI protocol to fetch packages with the `foo` namespace from ttl.sh/wasm-components
docs = { registry = "docs-registry",  metadata = { preferredProtocol = "oci", "oci" = {registry = "ttl.sh", namespacePrefix = "wasm-components/" } } }
```

> Note: the registry name can be referenced in the `package_registry_overrides` section of the `wkg` config to provide overrides for specific packages of a namespace.

## Distributing WIT and Components by Package Name with `wkg publish`

> Once you've [configured `wkg`](#wkg-configuration-files) to know where to publish packages to, you can use the `wkg publish` command to publish *components* or *interfaces* to be consumed by others.

Reviewer 2: It might be worth noting that you'll need to configure credentials for pushing and show an example. Can also just link to wkg docs again if easier!

Author: We reference that in the configuration section briefly:
```md
The configuration file also includes credentials for private registries, or for pushing to registries where you have permission, and other configuration options.  See the [`wkg` docs for more configuration options](https://github.com/bytecodealliance/wasm-pkg-tools?tab=readme-ov-file#configuration).
```

I think I'd prefer to leave that to the wkg documentation
